Möbius transform for arcs and lines

strobelm opened the pull request:
Hi after points here is the Möbius transform for Lines and Arcs. Feel free to comment on the pull request.

The last thing missing are conics, Cinderella treats them as circles (as far as i can tell), seems a bit odd to me. Any input on that?

gagern commented:
> The last thing missing are conics, Cinderella treats them as circles (as far as i can tell), seems a bit odd to me. Any input on that?

We shouldn't allow Möbius transformations of conics which are not circles, since the result will in general be quite complicated. Might revisit that if we ever have algebraic curves of higher degree as primitive elements, but not before. If you want to, you can try to work out what degree exactly we're talking about here, but in general I'd expect degree 4. That's because the Möbius transformation is already quadratic in the input coordinates, and plugging that into a quadratic form will lead to degree 4 unless by some lucky coincidence the highest degrees cancel out somehow.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1724,12 +1737,68 @@ geoOps.TrMoebiusP.kind = "P";
+    var l = csgeo.csnames[(el.args[1])].homog;
+
+    var getRandLine = function() {
+        var rline = List.realVector([100 * (Math.random() - 0.5), 100 * (Math.random() - 0.5), 100 * (Math.random() - 0.5)]);
```
Why the factor 100? I don't see that affecting the geometric meaning in any way, and I can't imagine it improving numerics either.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1724,12 +1737,68 @@ geoOps.TrMoebiusP.kind = "P";
+    var l1 = List.productMV(mat1, p);
+    var l2 = List.productMV(mat2, p);
+    return List.normalizeMax(List.cross(l1, l2));
+};
```
I'd suggest you pass in the Möbius transformation as a complete object, instead of passing the two matrices in separately. Everyone calling this code will be holding both in any case, so splitting things up only here means the rest of the code won't have to care about how a Möbius transformation is actually performed internally.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1724,12 +1737,68 @@ geoOps.TrMoebiusP.kind = "P";
+};
+
+geoOps.TrMoebiusArc = {};
+geoOps.TrMoebiusArc.kind = "C";
```
Can you try to get this hooked up to the Cinderella export via `geoMacros`?

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1211,19 +1211,32 @@ geoOps.CircleBy3 = {};
+    el.startPoint = csgeo.csnames[(el.args[0])].homog;
+    el.viaPoint = csgeo.csnames[(el.args[1])].homog;
+    el.endPoint = csgeo.csnames[(el.args[2])].homog;
+};
```
Why do we store this information with a circle that is not an arc? Seems like we shouldn't need it, and it might tempt some future contribution to use it inappropriately if we are not careful enough.

strobelm commented:
Here is the update: i fixed a bug in geoOps.RandomLine which prevented to have lines in the first sector (and removed the factor), added Moebius to geoMacros.Transform and prepared for further transformation and incorporated most of @gagern suggetions, thanks for that! 

Comments are very welcome.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1211,10 +1211,14 @@ geoOps.CircleBy3 = {};
     var d = List.jj;
     var p = csgeo.csnames[(el.args[2])].homog;
+    el.endPoint = p;
+
```
I'd still prefer to have these points not stored for a normal circle. Unless you have a good reason to do otherwise.

strobelm commented:
Just forgot about them, should be fixed now!

strobelm commented:
By the way Cinderella generates complete garbage when applying Möbius tranforms to conics, perhaps somebody with an account could open a bugreport (or even better create one for me ;-) ).

strobelm commented:
And here is the rebased version.

gagern commented:
> Cinderella generates complete garbage when applying Möbius tranforms to conics, perhaps somebody with an account could open a bugreport

Created issue 109 for this.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -7,8 +7,7 @@ geoOps._helper = {};
  * *s - Set of *
- * Tr - Projective transformation
- * Mt - Moebius transformation
+ * Tr - Transformation
```
Looks like something went wrong here during rebase. Or is this deliberate?

strobelm replied:
No it wasn't, reverted that.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1725,12 +1735,68 @@ geoOps.TrMoebiusP.kind = "P";
+    var Pts = geoOps._helper.TrMoebiusPoints(t, a1, a2, a3);
+    var b1 = Pts[0],
+        b2 = Pts[1],
+        b3 = Pts[2];
```
Makes me wonder about the use of `TrMoebiusPoints`. Why have a function which combines the three values to a single array, just to decompose that array again here? Right now, including the content of `TrMoebiusPoints` here would actually save you one line. Things would be different in ES2015 and its destructuring assignments, I guess.

strobelm replied:
You're right, i just wanted to avoid code doubling. Sometimes i really miss calling function by reference (without the object detour). I will upload a fix in a minute.

gagern commented on src/js/libgeo/GeoOps.js:
```diff
@@ -1928,14 +2018,23 @@ geoMacros.Calculation = function(el) {
+    var tr = csgeo.csnames[el.args[0]];
+    // workaround for Arcs since we treat them as circles
+    var akind = arg.kind;
+    if (arg.kind === "C" && arg.type === "ArcBy3") akind = "Arc";
```
Better use `arg.isArc` here, so you can repeatedly apply Möbius transformations to a segment or arc. I think you could drop the `kind === "C"` check and just check the `isArc` property alone.

Perhaps we should consider adding a kind for arcs, if and when we add a kind for circles. These things would be simpler if we had a type hierarchy here, so that we could add subtypes without having to change type checks in a million places. We might want to try the performance impact of such a change one day.

strobelm replied:
Good point, i think we need  ```arg.type === "ArcBy3" || arg.isArc``` since the macros are evaluated before initialization takes place.

I thought about introducing a new type, but since cinderella never had real use cases despite ArcBy3 i abandoned the idea.

gagern replied:
> … the macros are evaluated before initialization takes place.

Then how about moving the `isArc` property from the geometric element to the geometric operator? So you can do `geoOps[arg.type].isArc`. There definitely should be a test case for repeated transformations, since this is more tricky than it would seem at first glance.

By the way, #144 will make sure that every element gets fully initialized before any subsequent macro gets expanded.

strobelm replied:
Moving to geoOps isn't a bad idea, but we would have to keep the
el.isArc property since GeoRender.js will only have the element handed
over and can't distinguish circles and arcs anymore in case we remove this.

Perhaps our best shot is #144. I already read the code, but i still have
to run some examples.
 
On 12/12/2015 04:04 PM, Martin von Gagern wrote:
>>  geoMacros.Transform = function(el) {
>> -    var trKind = csgeo.csnames[el.args[0]].kind;
>> -    var argKind = csgeo.csnames[el.args[1]].kind;
>> +    var arg = csgeo.csnames[el.args[1]];
>> +    var tr = csgeo.csnames[el.args[0]];
>> +    // workaround for Arcs since we treat them as circles
>> +    var akind = arg.kind;
>> +    if (arg.kind === "C" && arg.type === "ArcBy3") akind = "Arc";
>> … the macros are evaluated before initialization takes place.
> Then how about moving the `isArc` property from the geometric element to the geometric operator? So you can do `geoOps[arg.type].isArc`. There definitely should be a test case for repeated transformations, since this is more tricky than it would seem at first glance.
>
> By the way, #144 will make sure that every element gets fully initialized before any subsequent macro gets expanded.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/CindyJS/CindyJS/pull/141/files#r47432651


-- 
Michael Strobel, M.Sc.

Technische Universität München -- Zentrum Mathematik
Lehrstuhl für Geometrie und Visualisierung (M10)
Boltzmannstr. 3, 85747 Garching bei München 
https://www-m10.ma.tum.de/bin/view/Lehrstuhl/MichaelStrobel

gagern replied:
You could easily use the same check using `geoOps` over in ` GeoRender.js` as well. But if this is a motivation to go for #144, I'll be more than happy to see that proceed.

strobelm commented:
Here is the rebased version with create merged. Since the initialization problem is gone it's easier to detect arcs.

elkins0 commented:
Alone these appear to work fine. However with TrInverse the geometry won't load:

```javascript
geoOps.TrInverse = {};
geoOps.TrInverse.kind = "Tr";
geoOps.TrInverse.updatePosition = function(el) {
    var tr = csgeo.csnames[(el.args[0])];
    var m = tr.matrix;
    el.dualMatrix = List.transpose(tr.matrix); <--
    el.matrix = List.transpose(tr.dualMatrix);
};

List.transpose = function(a) {
    var erg = [];
    var n = a.value[0].value.length; <--
Unable to get property 'value' of undefined or null reference
```
You have setup mat1 and mat2. However matrix and dualMatrix are undefined.

strobelm commented:
Hi,

yes this is true, simply because inversion for Möbiustransformations
isn't that easy. For inverting a transform (az+b)/(cz+d) we have to
compute (dz -b)/(-cz+a) and therefor recalculate mat1 and mat2. This
needs it's own function.


Am 15.12.2015 um 22:43 schrieb Alexander Elkins:
>
> Alone these appear to work fine. However with TrInverse the geometry
> won't load:
>
> geoOps.TrInverse = {};
> geoOps.TrInverse.kind = "Tr";
> geoOps.TrInverse.updatePosition = function(el) {
>     var tr = csgeo.csnames[(el.args[0])];
>     var m = tr.matrix;
>     el.dualMatrix = List.transpose(tr.matrix); <--
>     el.matrix = List.transpose(tr.dualMatrix);
> };
>
> List.transpose = function(a) {
>     var erg = [];
>     var n = a.value[0].value.length; <--
> Unable to get property 'value' of undefined or null reference
>
> You have setup mat1 and mat2. However matrix and dualMatrix are undefined.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CindyJS/CindyJS/pull/141#issuecomment-164906234>.
>

-- 
Michael Strobel, M.Sc.

Technische Universität München -- Zentrum Mathematik
Lehrstuhl für Geometrie und Visualisierung (M10)
Boltzmannstr. 3, 85747 Garching bei München 
https://www-m10.ma.tum.de/bin/view/Lehrstuhl/MichaelStrobel

gagern commented:
We should also discuss how to handle this from Cinderella. The Cinderella representation has a tendency to use the same name for different but related operations. Like `Transform` for transforming “an object”, no matter what object that is, or what transformation. But also `TrInverse` which it apparently uses to invert both projective and Möbius transformations. The rub here is that we are already using that name, exclusively for projective transformations.

So I think we have these options:

1. Change the name Cinderella exports. We have name-changing machinery in place, and we at some point decided to use the more descriptive names, which in this case and in my opinion means differentiating by type of transformation. Should we change other names as well, to cover the full set of CindyJS transformation primitives? I think it would be consistent to do so.

2. Change the name of what's now called `TrInverse` (to something like `ProjectiveInverse` or whatever) so that we can make `TrInverse` a macro. The renaming is neccessary since macro expansion is deliberately recursive, so we can't have the same name for a macro and a primitive.

3. Write some special case code. I'd rather avoid this unless there is some *very* good reason.

I think I'd go with 1. @kortenkamp, what's your opinion here?

gagern commented on examples/111_Moebius_CinderellaExportTest.html:
```diff
@@ -0,0 +1,61 @@
+		{ name: "U", type: "Transform", pos: [ 4.0, 3.155084971423055, 0.1144918446644713 ], color: [ 1.0, 0.0, 0.0 ], args: [ "Tr1", "H" ], labeled: true }, 
+		{ name: "V", type: "Transform", pos: [ 4.0, 2.926257936039025, 0.12327138542933927 ], color: [ 1.0, 0.0, 0.0 ], args: [ "Tr1", "K" ], labeled: true } ], 
+	ports: [ 
+		{ id: "CSCanvas", width: 3200, height: 1548, transform: [ { visibleRect: [ -24.44, 32.4, 103.56, -29.52 ] } ], background: "rgb(168,176,192)" } ] });
```
Those dimensions are pretty hard to use for mere mortals who don't run a HiDPI-screen at native resolution…

Seriously, if you have that Cinderella file lying around, it would be nice to see it re-exported at a lower zoom and window size. Otherwise we might have to adjust the viewport manually to something more reasonable. `111_Moebius_It` is a bit on the large side as well.

strobelm commented:
Your wish is my command ;-). I updated the examples to a more SD friendly resolution, the iterated one shows nicely the drawing bug  #139 which i will tackle as soon this gets merged.

gagern commented:
Yes, I've seen #139 in both the updated examples. But that won't keep the merge here. Thanks for all the work!

gagern commented:
Looking at #24, `TrMoebiusS` seems to be still missing, right?
